code/modules/atmospherics/machinery/components/binary_devices/thermomachine.dm

```diff
@@ -0,0 +1,367 @@
+		calculated_bin_rating += bin.rating
+	heat_capacity = 5000 * ((calculated_bin_rating - 1) ** 2)
+	min_temperature = T20C
+	max_temperature = T20C
+	if(cooling)
+		var/calculated_laser_rating
+		for(var/obj/item/stock_parts/micro_laser/laser in component_parts)
+			calculated_laser_rating += laser.rating
+		min_temperature = max(T0C - (base_cooling + calculated_laser_rating * 15), TCMB) //73.15K with T1 stock parts
+	else
+		var/calculated_laser_rating
+		for(var/obj/item/stock_parts/micro_laser/laser in component_parts)
+			calculated_laser_rating += laser.rating
+		max_temperature = T20C + (base_heating * calculated_laser_rating) //573.15K with T1 stock parts
+
+/obj/machinery/atmospherics/components/binary/thermomachine/update_icon_state()
+	if(panel_open)
+		icon_state = icon_state_open
+		return ..()
+	if(on && is_operational)
+		icon_state = icon_state_on
+		return ..()
+	icon_state = icon_state_off
+	return ..()
+
+/obj/machinery/atmospherics/components/binary/thermomachine/update_overlays()
+	. = ..()
+	. += getpipeimage(icon, "pipe", dir, COLOR_LIME, piping_layer)
+	. += getpipeimage(icon, "pipe", turn(dir, 180), COLOR_MOSTLY_PURE_RED, piping_layer)
+	if(holding)
+		var/mutable_appearance/holding = mutable_appearance(icon, "holding")
+		. += holding
+
+/obj/machinery/atmospherics/components/binary/thermomachine/update_icon_nopipes()
+	cut_overlays()
+	if(showpipe)
+		add_overlay(getpipeimage(icon, "scrub_cap", initialize_directions))
+
+/obj/machinery/atmospherics/components/binary/thermomachine/examine(mob/user)
+	. = ..()
+	. += "<span class='notice'>The thermostat is set to [target_temperature]K ([(T0C-target_temperature)*-1]C).</span>"
+	if(in_range(user, src) || isobserver(user))
+		. += "<span class='notice'>The status display reads: Efficiency <b>[(heat_capacity/5000)*100]%</b>.</span>"
+		. += "<span class='notice'>Temperature range <b>[min_temperature]K - [max_temperature]K ([(T0C-min_temperature)*-1]C - [(T0C-max_temperature)*-1]C)</b>.</span>"
+
+/obj/machinery/atmospherics/components/binary/thermomachine/AltClick(mob/living/user)
+	if(!can_interact(user))
+		return
+	if(cooling)
+		target_temperature = min_temperature
+		investigate_log("was set to [target_temperature] K by [key_name(user)]", INVESTIGATE_ATMOS)
+		to_chat(user, "<span class='notice'>You minimize the target temperature on [src] to [target_temperature] K.</span>")
+	else
+		target_temperature = max_temperature
+		investigate_log("was set to [target_temperature] K by [key_name(user)]", INVESTIGATE_ATMOS)
+		to_chat(user, "<span class='notice'>You maximize the target temperature on [src] to [target_temperature] K.</span>")
+
+/obj/machinery/atmospherics/components/binary/thermomachine/process_atmos()
+	if(!is_operational || !on)  //if it has no power or its switched off, dont process atmos
+		on = FALSE
+		update_appearance()
+		return
+
+	var/turf/local_turf = get_turf(src)
+	var/datum/gas_mixture/enviroment = local_turf.return_air()
+	if(!local_turf)
```

Comment:
If this is true, you would have already thrown an exception on the previous line

Reply:
moved above